src/librustc_privacy/lib.rs

```diff
@@ -20,22 +20,260 @@ use rustc::hir::intravisit::{self, Visitor, NestedVisitorMap};
+                    return def_id_visibility(tcx, tcx.hir().get_parent_did(node_id));
+                }
+                Node::ImplItem(impl_item) => {
+                    match tcx.hir().get(tcx.hir().get_parent(node_id)) {
+                        Node::Item(item) => match &item.node {
+                            hir::ItemKind::Impl(.., None, _, _) => &impl_item.vis,
+                            hir::ItemKind::Impl(.., Some(trait_ref), _, _)
+                                => return def_id_visibility(tcx, trait_ref.path.def.def_id()),
+                            kind => bug!("unexpected item kind: {:?}", kind),
+                        }
+                        node => bug!("unexpected node kind: {:?}", node),
+                    }
+                }
+                Node::StructCtor(vdata) => {
+                    let struct_node_id = tcx.hir().get_parent(node_id);
+                    let item = match tcx.hir().get(struct_node_id) {
+                        Node::Item(item) => item,
+                        node => bug!("unexpected node kind: {:?}", node),
+                    };
+                    let (mut ctor_vis, mut span, mut descr) =
+                        (ty::Visibility::from_hir(&item.vis, struct_node_id, tcx),
+                         item.vis.span, item.vis.node.descr());
+                    for field in vdata.fields() {
+                        let field_vis = ty::Visibility::from_hir(&field.vis, node_id, tcx);
+                        if ctor_vis.is_at_least(field_vis, tcx) {
+                            ctor_vis = field_vis;
+                            span = field.vis.span;
+                            descr = field.vis.node.descr();
+                        }
+                    }
+
+                    // If the structure is marked as non_exhaustive then lower the
+                    // visibility to within the crate.
+                    if ctor_vis == ty::Visibility::Public {
+                        let adt_def = tcx.adt_def(tcx.hir().get_parent_did(node_id));
+                        if adt_def.non_enum_variant().is_field_list_non_exhaustive() {
+                            ctor_vis = ty::Visibility::Restricted(DefId::local(CRATE_DEF_INDEX));
+                            span = attr::find_by_name(&item.attrs, "non_exhaustive").unwrap().span;
+                            descr = "crate-visible";
+                        }
+                    }
+
+                    return (ctor_vis, span, descr);
+                }
+                Node::Expr(expr) => {
+                    return (ty::Visibility::Restricted(tcx.hir().get_module_parent(expr.id)),
+                            expr.span, "private")
+                }
+                node => bug!("unexpected node kind: {:?}", node)
+            };
+            (ty::Visibility::from_hir(vis, node_id, tcx), vis.span, vis.node.descr())
+        }
+        None => (tcx.visibility(def_id), tcx.def_span(def_id), "private"),
```

Comment:
Aren't items from other crates public? i.e., why is this the string "private"?

Reply:
This is me being lazy, the returned string is never used if the def-id is from another crate.
Will fix.

Reply:
Perhaps I need to just return `"<unreachable>"` :)

Also, items from other crates can be private in erroneous code.